diff --git a/arch/arm/boot/dts/meson8b-mxq.dts b/arch/arm/boot/dts/meson8b-mxq.dts
index f3937d55472d..4fc1565ac87a 100644
--- a/arch/arm/boot/dts/meson8b-mxq.dts
+++ b/arch/arm/boot/dts/meson8b-mxq.dts
@@ -32,7 +32,7 @@
 
 		regulator-name = "VCCK";
 		regulator-min-microvolt = <860000>;
-		regulator-max-microvolt = <1140000>;
+		regulator-max-microvolt = <1160000>;
 
 		pwms = <&pwm_cd 0 1148 0>;
 		pwm-dutycycle-range = <100 0>;
@@ -93,6 +93,13 @@
 	cpu-supply = <&vcck>;
 };
 
+// minimal s805 overclocking
+&cpu_opp_table {
+	opp-1632000000 {
+		status = "okay";
+	};
+};
+
 &ethmac {
 	status = "okay";
 
diff --git a/arch/arm/boot/dts/meson8b.dtsi b/arch/arm/boot/dts/meson8b.dtsi
index 01cd5a885e5e..bc5f5cabf855 100644
--- a/arch/arm/boot/dts/meson8b.dtsi
+++ b/arch/arm/boot/dts/meson8b.dtsi
@@ -123,6 +123,15 @@
 			opp-hz = /bits/ 64 <1536000000>;
 			opp-microvolt = <1140000>;
 		};
+// minimal s805 overclocking - higher frequencies than
+// this do not result in higher frequencies in reality
+// see: https://github.com/hardkernel/linux/commit/594d995163d18449a48cd4e06b989c6de1493658
+// and: https://nixdev.com/?p=397
+		opp-1632000000 {
+			opp-hz = /bits/ 64 <1632000000>;
+			opp-microvolt = <1160000>;
+			status = "disabled";
+		};
 	};
 
 	gpu_opp_table: gpu-opp-table {
